### PR TITLE
add exception for the test

### PR DESCRIPTION
**Pull request by hchoura**
To catch exception in the rule engine and in the contextAgent. 

It will cause several exception for other project.

### CODE REVIEW

File: src/foam/core/CompoundContextAgency.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ for ( Runnable agent : getAgents() ) {
   } catch (Throwable t) {
     logger.error(t.getMessage(), t);
     e.add(t);
+    throw t;
```
Comment (jlhughes) [inline src/foam/core/CompoundContextAgency.js]: @hchoura the throwable is added to the compoundexception. 
You can't throw it.  This is nonsense.

**Review by jlhughes — CHANGES_REQUESTED** [review]

File: src/foam/core/CompoundContextAgency.js (hunk 1 above)
Comment (kgrgreer) [reply src/foam/core/CompoundContextAgency.js]: Yes, this doesn't make sense. If there is an exception, then we'll rethrow it on line 45 below:
e.maybeThrow();

File: src/foam/nanos/ruler/RuleEngine.java [hunk 2]
```diff
@@ -84,6 +84,7 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
       compoundAgency.execute(x_);
     } catch (Exception e) {
       logger.error(e.getMessage(), e);
+      throw e;
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RuleEngine.java:87]: this is occuring asynchronously, there is nothing to catch the exception. Reporting is all we can do.  Perhaps we can add it to the rule - so a report of what failed. but we can't throw from here.

File: src/foam/core/CompoundContextAgency.js [hunk 3]
```diff
@@ -42,7 +42,11 @@ for ( Runnable agent : getAgents() ) {
     e.add(t);
   }
 }
-e.maybeThrow();`
+try {
+  e.maybeThrow();
+} catch ( Throwable e1 ) {
+  e1.printStackTrace();
+}`
```
Comment (jlhughes) [inline src/foam/core/CompoundContextAgency.js]: if e.maybeThrow() throws then the e1 needs to be rethrown.

File: src/foam/core/CompoundException.java [hunk 4]
```diff
@@ -11,15 +11,16 @@
...
   public void add(Throwable t) {
     exceptions_.add(t);
   }
 
-  public void maybeThrow() {
-    if ( exceptions_.size() != 0 ) {
-      throw this;
+  public void maybeThrow() throws Throwable {
+    for ( int i = 0 ; i < exceptions_.size() ; i++ ) {
+      System.out.println( "This is an error =======> " + ( (Throwable) exceptions_.get(i) ) );
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]: We only want this in TEST mode.

File: src/foam/core/CompoundException.java [hunk 5]
```diff
@@ -11,15 +11,16 @@
...
+      throw (Throwable) exceptions_.get(i);
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]: and why are we throwing on the first?

File: src/foam/core/ContextAgentRunnable.java [hunk 6]
```diff
@@ -24,6 +24,11 @@ public String toString() {
   }
 
   public void run() {
-    agent_.execute(x_);
+    try {
+      agent_.execute(x_);
+    } catch ( Throwable e ) {
+      e.printStackTrace();
```
Comment (jlhughes) [inline src/foam/core/ContextAgentRunnable.java]: Again only in TEST mode.

File: src/foam/core/ContextAgentRunnable.java [hunk 7]
```diff
@@ -24,6 +24,11 @@ public String toString() {
...
+      throw e;
```
Comment (jlhughes) [inline src/foam/core/ContextAgentRunnable.java]: and this is pointless. Who is catching it?

File: src/foam/core/CompoundContextAgency.js [hunk 8]
```diff
@@ -42,7 +42,11 @@ for ( Runnable agent : getAgents() ) {
     e.add(t);
   }
 }
-e.maybeThrow();`
+try {
```
Comment (kgrgreer) [inline src/foam/core/CompoundContextAgency.js]: What's the point of throwing the exception if we're just going to catch it.

File: src/foam/core/CompoundException.java (hunk 5 above)
Comment (kgrgreer) [inline src/foam/core/CompoundException.java]: This will only throw the first exception. We want to throw them all.

File: src/foam/core/ContextAgentRunnable.java (hunk 6 above)
Comment (kgrgreer) [inline src/foam/core/ContextAgentRunnable.java]: Why is this needed? This will just pollute the stdout.

File: src/foam/core/CompoundContextAgency.js [hunk 9]
```diff
@@ -42,7 +43,8 @@ for ( Runnable agent : getAgents() ) {
     e.add(t);
   }
 }
-e.maybeThrow();`
+e.maybeThrow((AppConfig) getX().get("appConfig"));
```
Comment (jlhughes) [inline src/foam/core/CompoundContextAgency.js:46]: How about passing a context - that would be consistent with the rest of FOAM.

File: src/foam/core/CompoundException.java [hunk 10]
```diff
@@ -7,19 +7,24 @@
...
     exceptions_.add(t);
   }
 
-  public void maybeThrow() {
-    if ( exceptions_.size() != 0 ) {
-      throw this;
+  public void maybeThrow(AppConfig appConfig) {
+    if ( Mode.TEST == appConfig.getMode() ) {
+      for ( int i = 0 ; i < exceptions_.size() ; i++ ) {
+        System.out.println( "This is an error =======> " + ( (Throwable) exceptions_.get(i) ) );
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java:25]: Please implement toString on CompoundException which can generate an appropriate message.

File: src/foam/core/CompoundException.java [hunk 11]
```diff
@@ -7,19 +7,24 @@
...
+        throw (RuntimeException) exceptions_.get(i);
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java:26]: and this still makes no sense.

**Review by jlhughes — CHANGES_REQUESTED** [review]

**Comment by jlhughes** [comment]
Superceded by another PR.
